(dev/core#174) Full PSR-16 compliance for ArrayCache, SqlGroup, Redis, Memcache, APC

totten opened the pull request:
Overview
----------------------------------------
#12342 provides nominal compliance with PSR-16; however, some drivers raise exceptions or warnings for new options . This PR ensures more complete support for PSR-16 in three drivers (`ArrayCache`, `SqlGroup`, and `Redis`).

(~~__Depends__: The patches here work/are cogent, but I've flagged it "Blocked" because it depends/includes other open PRs: #12342, #12348, #12354~~)

Before
----------------------------------------
* `ArrayCache`, `SqlGroup`, `Redis`, `Memcache`, `Memcached`, and `APCcache` are not tested for compliance -- and would not pass if they were because:
    * `$cache->set(...$ttl)` would throw an error any custom TTL.
    * `$cache->get(...$default)` woudl throw an error for any not-NULL default.
    * Cache-keys are not validated.
    * Some variants of `$cache->set()` / `$cache->get()` handle objects in ways that are unsafe.
    * In some variants, `$cache->flush()` is overzealous (Memcache); in others, underzealous (APCcache).

After
----------------------------------------
* `ArrayCache`, `SqlGroup`, `Redis`, `Memcache`, `Memcached`, and `APCcache` have been fixed to comply (per [SimpleCacheTest](https://github.com/civicrm/civicrm-packages/pull/215/files)), and they comply.

Technical Details
----------------------------------------
* `SqlGroup` is substantially rewritten. It no longer wraps around `CRM_Core_BAO_Cache` - instead, it does SQL queries which are similar in nature and uses its own front-cache.
    * The interface and implementation of `CRM_Core_BAO_Cache` do not support TTL/expiration. Adding support effectively requires patching every function to respect TTL in reading+writing both DB and thread-local front-cache. And in the long-run, it's better to maintain that kind of logic in framework that is standard compliant.
    * I was concerned about a theoretical issue -- if `CRM_Core_BAO_Cache` and `CRM_Utils_Cache_SqlGroup` wrap the same DB table but have different front-caches, and if you tried accessing the same records through both interfaces, then there could be cache-coherence bugs. However, if we consider [the audit of caches](https://lab.civicrm.org/dev/core/issues/174#note_5812), it becomes less of a concern. Why? Because the use-cases are XOR. (Ex: The `dashboard` cache always uses BAO to read/write cache-rows. Ex: The `js_strings` cache always uses OOP to read/write cache-rows.)
    * At some point, we may want to fully flip the relationship around (e.g. `BAO_Cache` could just be a thin wrapper for `SqlGroup` which we provide for backward compatibility). However, that will make even bigger PR that's harder to review en masse.
* The test coverage for `SqlGroup` and `ArrayCache` should work intuitively. For `Redis`, `Memcache`, and `Memcached`, the test needs a particular environment to run -- which isn't universally available. The `ConfiguredMemoryTest` will run against whatever cache service is enabled in `civicrm.settings.php` -- and if no service is configured, it will be skipped. (`APCcache` is similar, although the specific environmental requirements differ.) For the tests which don't run in the CI environment, I've confirmed they pass locally.

seamuslee001 commented on CRM/Utils/Cache/SqlGroup.php:
```diff
@@ -89,22 +107,82 @@ public function __construct($config) {
+    //    CRM_Core_BAO_Cache::setItem($value, $this->group, $key, $this->componentID);
+    $lock = Civi::lockManager()->acquire("cache.{$this->group}_{$key}._null");
+    if (!$lock->isAcquired()) {
+      CRM_Core_Error::fatal();
```
@eileenmcnaughton @totten should we throw exception here rather than fatal?

totten replied:
@seamuslee001 Probably. Thinking out loud:

* The system is generally simpler/clearer with exceptions.
* If there's anyone using `SqlGroup` who's trying to catch/handle problems, they could be surprised by a switch from `fatal` to `exception`.
* OTOH, very few things try to handle fatals. (Because they're awkward to handle.) And when they are handled, that's by converting the fatal to an exception. (Not strictly so - but in practical terms, that's all I've ever seen or heard of.)
* Overall, the previous contract for `CRM_Utils_Cache_Interface` was unclear about how errors would be reported. PSR-16 pushes us to standardize (i.e. returning booleans and throwing particular exception-interfaces). 
* I suspect real-world error-handling behaviors in `CRM/Utils/Cache/*` have been pretty varied (depending on the particular driver/configuration, you might get exceptions or fatals or FALSE return values). Any problem that rises from "making the error handling more consistent" is likely to be in the category of "the problem existed before but manifested differently".
* We don't often get to highlight the cache-subsystem as an area of change. It may be net-gentler on downstream to rip-off this particular bandaid rather than kick it down the road.

So... yes, I'll update to use the exception instead.

totten commented:
jenkins, test this please

totten commented:
Rebased with some updates:
* Fixed typo in various commit messages (173 vs 174)
* Added _(dev/core#174) CRM_Utils_Cache::assertValidKey - Relax enforcement at runtime_ (d6d6320c0b). This doesn't replace #12348, but it should fail more gracefully in production if there are oddball scenarios that we haven't found/fixed.

seamuslee001 commented:
@totten are you able to rebase this now please?

seamuslee001 commented on CRM/Utils/Cache/Memcached.php:
```diff
@@ -110,29 +127,79 @@ public function __construct($config) {
+        throw new CRM_Utils_Cache_CacheException("Memcached::set($key) failed: " . $this->_cache->getResultMessage());
+      }
+      else {
+        Civi::log()->error("Redis set ($key) failed: " . $this->_cache->getResultMessage());
```
@totten i think this should be Memcached here

totten replied:
Thanks. Fixed and squashed.

eileenmcnaughton commented:
I wish we had stats on the use of these - my guess is it would be better to move some out to an extension at some point

eileenmcnaughton commented on CRM/Utils/Cache.php:
```diff
@@ -214,4 +214,30 @@ public static function create($params = array()) {
+   * @throws \CRM_Utils_Cache_InvalidArgumentException
+   */
+  public static function assertValidKey($key) {
+    $strict = CRM_Utils_Constant::value('CIVICRM_PSR16_STRICT', FALSE) || defined('CIVICRM_TEST');
```
I like this

eileenmcnaughton commented on CRM/Utils/Cache/APCcache.php:
```diff
@@ -97,10 +101,12 @@ public function set($key, $value, $ttl = NULL) {
-    if ($default !== NULL) {
-      throw new \RuntimeException("FIXME: " . __CLASS__ . "::get() only supports NULL default");
+    CRM_Utils_Cache::assertValidKey($key);
+    $result = apc_fetch($this->_prefix . $key, $success);
```
$success does not seem to be defined

totten replied:
That sounds like some metaphysical management widsom. ;)

More seriously, `$success` is actually an output of this line. [apc_fetch($key, &$success)](http://php.net/apc_fetch) sets the value of `$success` to indicate whether it worked.

eileenmcnaughton commented on CRM/Utils/Cache/NaiveMultipleTrait.php:
```diff
@@ -92,11 +99,23 @@ public function setMultiple($values, $ttl = NULL) {
+   * @throws \CRM_Utils_Cache_InvalidArgumentException
+   */
+  private function assertIterable($func, $keys) {
+    if (!is_array($keys) || $keys instanceof Traversable) {
```
I feel like this should be reversed ie

```
While you cannot implement this interface, you can use it in your checks to determine if something is usable in for each. Here is what I use if I'm expecting something that must be iterable via foreach.
in the example not !items
```
<?php
    if( !is_array( $items ) && !$items instanceof Traversable )
        //Throw exception here
?>
```
http://php.net/manual/en/class.traversable.php

totten replied:
Oops, good catch! This explains why `testGetMultipleWithGenerator()`,`testSetMultipleWithGenerator()`, and `testDeleteMultipleGenerator()` weren't working. :)

I've pushed up a patch to fix `assertIterable` and re-enable those tests.

totten commented:
Most of this has been split-out to smaller PRs. There was one last bit, which I've now put in #12389, so we'll close this one.
